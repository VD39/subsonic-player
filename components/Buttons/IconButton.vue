<script setup lang="ts">
const props = withDefaults(
  defineProps<{
    disabled?: boolean;
    fullWidth?: boolean;
    icon?: string;
    iconPosition?: IconPosition;
    iconSize?: number;
    is?: ButtonType;
    showText?: boolean;
    type?: string;
  }>(),
  {
    disabled: undefined,
    icon: undefined,
    iconPosition: 'left',
    iconSize: 20,
    is: 'button',
    type: 'button',
  },
);

const type = props.is === 'button' ? props.type : undefined;
const isComponent =
  props.is === 'nuxt-link' ? resolveComponent('NuxtLink') : props.is;
</script>

<template>
  <component
    :is="isComponent"
    :class="[
      $style.iconButton,
      {
        [$style.alignRight]: icon && iconPosition === 'right',
        [$style.disabled]: disabled,
        [$style.fullWidth]: fullWidth,
      },
    ]"
    :disabled="disabled"
    :type="type"
    v-bind="$attrs"
  >
    <component
      :is="icon"
      v-if="icon"
      ref="phIcon"
      :class="$style.icon"
      aria-hidden="true"
      :size="iconSize"
      weight="duotone"
    />

    <span
      ref="text"
      :class="[
        $style.text,
        {
          'visually-hidden': !showText,
        },
      ]"
    >
      <slot />
    </span>
  </component>
</template>

<style module>
.iconButton {
  @mixin align-center;

  position: relative;
  gap: var(--space-8);
  padding: var(--space-8);
  margin: 0;
  white-space: nowrap;
}

.alignRight {
  flex-direction: row-reverse;
}

.fullWidth {
  width: 100%;
}

.disabled {
  pointer-events: none;
  opacity: 0.25;
}

.icon {
  fill: var(--body-font-color);

  [opacity='0.2'] {
    opacity: 0.4;
    fill: var(--secondary-font-color);
  }
}

.text {
  display: block;
  margin-bottom: 1px;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}
</style>
