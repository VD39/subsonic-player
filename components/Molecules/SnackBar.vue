<script setup lang="ts">
import ButtonLink from '@/components/Atoms/ButtonLink.vue';
import MessageBar from '@/components/Atoms/MessageBar.vue';

const { removeSnack, snacks } = useSnack();
</script>

<template>
  <Teleport to="#teleports">
    <transition name="slide-left-right">
      <div v-if="snacks.length" ref="snackbarWrapper" :class="$style.wrapper">
        <transition-group name="slide-left-right">
          <div
            v-for="snack in snacks"
            :key="snack.id"
            class="mBM"
            :class="$style.snackbar"
          >
            <MessageBar :type="snack.type">
              <p ref="content" class="sentenceCase">
                {{ snack.content }}
              </p>

              <ButtonLink :icon="ICONS.close" @click="removeSnack(snack.id)">
                Clear snack
              </ButtonLink>
            </MessageBar>
          </div>
        </transition-group>
      </div>
    </transition>
  </Teleport>
</template>

<style module>
.wrapper {
  position: fixed;
  inset: auto 0 0;
  z-index: 9;
  margin: var(--space-16) var(--space-16) 0;

  @media (--tablet-up) {
    left: auto;
  }
}

.snackbar {
  @media (--tablet-up) {
    width: min(50vw, 350px);
  }
}
</style>
