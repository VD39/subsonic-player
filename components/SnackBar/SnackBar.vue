<script setup lang="ts">
import IconButton from '@/components/Buttons/IconButton.vue';

const { snacks, removeSnack } = useSnack();
</script>

<template>
  <Teleport to="#teleports">
    <transition name="slide-left-right">
      <div v-if="snacks.length" ref="snackbarWrapper" :class="$style.wrapper">
        <transition-group name="slide-left-right">
          <div v-for="snack in snacks" :key="snack.id">
            <div
              data-test-id="content-wrapper"
              :class="[$style[snack.type], $style.snackbar]"
            >
              <p data-test-id="content" :class="$style.content">
                {{ snack.content }}
              </p>

              <IconButton icon="PhX" @click="removeSnack(snack.id)">
                Clear snack
              </IconButton>
            </div>
          </div>
        </transition-group>
      </div>
    </transition>
  </Teleport>
</template>

<style module>
.wrapper {
  position: fixed;
  inset: auto 0 0 auto;
  z-index: 5;
  margin: var(--space-16) var(--space-16) 0;
}

.snackbar {
  @mixin align-center;

  gap: var(--space-8);
  justify-content: space-between;
  width: min(50vw, 350px);
  padding: var(--space-16);
  margin-bottom: var(--space-8);
  font-weight: 600;
  border-radius: var(--border-radius-medium);
}

.success {
  background-color: var(--success-color);
}

.error {
  background-color: var(--error-color);
}

.info {
  background-color: var(--info-color);
}

.content {
  font-size: var(--small-font-size);
  word-break: break-word;
  word-wrap: break-word;
  overflow-wrap: break-word;
}
</style>
