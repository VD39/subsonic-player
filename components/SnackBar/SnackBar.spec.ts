import type { VueWrapper } from '@vue/test-utils';
import { mount } from '@vue/test-utils';
import { mockNuxtImport } from '@nuxt/test-utils/runtime';
import IconButton from '@/components/Buttons/IconButton.vue';
import SnackBar from './SnackBar.vue';

const snacksMock = ref<Snack[]>([]);

mockNuxtImport('useSnack', () => () => ({
  snacks: snacksMock,
  removeSnack: vi.fn((id: string) => {
    snacksMock.value = snacksMock.value.filter((snack) => snack.id !== id);
  }),
}));

function factory(props = {}) {
  return mount(SnackBar, {
    props: {
      ...props,
    },
    global: {
      stubs: {
        teleport: true,
      },
    },
  });
}

describe('SnackBar', () => {
  let wrapper: VueWrapper;

  beforeEach(() => {
    wrapper = factory();
  });

  describe('when snacks values are not set', () => {
    it('matches the snapshot', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('does not show the snackbar wrapper', () => {
      expect(wrapper.find({ ref: 'snackbarWrapper' }).exists()).toBe(false);
    });
  });

  describe('when snacks values are set', () => {
    describe.each([
      [
        {
          id: 'id',
          content: 'content',
          type: 'error',
        },
      ],
      [
        {
          id: 'id',
          content: 'content',
          type: 'success',
        },
      ],
      [
        {
          id: 'id',
          content: 'content',
          type: 'info',
        },
      ],
    ])('when value is %o', (value) => {
      beforeEach(() => {
        snacksMock.value = [value as Snack];
      });

      afterEach(() => {
        snacksMock.value = [];
      });

      it('matches the snapshot', () => {
        expect(wrapper.html()).toMatchSnapshot();
      });

      it('shows the snackbar wrapper', () => {
        expect(wrapper.find({ ref: 'snackbarWrapper' }).exists()).toBe(true);
      });

      it('adds the correct class to content wrapper', () => {
        expect(
          wrapper.find('[data-test-id="content-wrapper"]').classes(),
        ).toContain(value.type);
      });

      it('sets the correct content', () => {
        expect(wrapper.find('[data-test-id="content"]').text()).toBe(
          value.content,
        );
      });
    });

    describe('when snacks values are more than 1', () => {
      beforeEach(() => {
        snacksMock.value = Array(4)
          .fill('')
          .map((_item, index) => ({
            id: `snack-${index}`,
            content: `content-${index}`,
            type: 'error',
          }));
      });

      it('matches the snapshot', () => {
        expect(wrapper.html()).toMatchSnapshot();
      });

      it('shows the correct amount of snack content', () => {
        expect(wrapper.findAll('[data-test-id="content-wrapper"]').length).toBe(
          4,
        );
      });

      describe('when remove button is clicked', () => {
        beforeEach(() => {
          wrapper.findAllComponents(IconButton)[0].vm.$emit('click');
        });

        it('shows the correct amount of snack content', () => {
          expect(
            wrapper.findAll('[data-test-id="content-wrapper"]').length,
          ).toBe(3);
        });
      });
    });
  });
});
