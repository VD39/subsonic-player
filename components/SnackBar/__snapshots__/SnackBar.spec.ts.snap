// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SnackBar > when snacks values are not set > matches the snapshot 1`] = `
"<transition-stub name="snackbar" appear="false" persisted="false" css="true">
  <!--v-if-->
</transition-stub>"
`;

exports[`SnackBar > when snacks values are set > when snacks values are more than 1 > matches the snapshot 1`] = `
"<transition-stub name="snackbar" appear="false" persisted="false" css="true">
  <div class="wrapper">
    <transition-group-stub name="snackbar" appear="false" persisted="false" css="true">
      <div>
        <div data-test-id="content-wrapper" class="error snackbar">
          <p data-test-id="content" class="content">content-0</p><button class="iconButton" type="button">
            <phxcircle class="icon" aria-hidden="true" size="20" weight="duotone"></phxcircle><span class="text visually-hidden"> Clear snack </span>
          </button>
        </div>
      </div>
      <div>
        <div data-test-id="content-wrapper" class="error snackbar">
          <p data-test-id="content" class="content">content-1</p><button class="iconButton" type="button">
            <phxcircle class="icon" aria-hidden="true" size="20" weight="duotone"></phxcircle><span class="text visually-hidden"> Clear snack </span>
          </button>
        </div>
      </div>
      <div>
        <div data-test-id="content-wrapper" class="error snackbar">
          <p data-test-id="content" class="content">content-2</p><button class="iconButton" type="button">
            <phxcircle class="icon" aria-hidden="true" size="20" weight="duotone"></phxcircle><span class="text visually-hidden"> Clear snack </span>
          </button>
        </div>
      </div>
      <div>
        <div data-test-id="content-wrapper" class="error snackbar">
          <p data-test-id="content" class="content">content-3</p><button class="iconButton" type="button">
            <phxcircle class="icon" aria-hidden="true" size="20" weight="duotone"></phxcircle><span class="text visually-hidden"> Clear snack </span>
          </button>
        </div>
      </div>
    </transition-group-stub>
  </div>
</transition-stub>"
`;

exports[`SnackBar > when snacks values are set > when value is { id: 'id', content: 'content', type: 'error' } > matches the snapshot 1`] = `
"<transition-stub name="snackbar" appear="false" persisted="false" css="true">
  <div class="wrapper">
    <transition-group-stub name="snackbar" appear="false" persisted="false" css="true">
      <div>
        <div data-test-id="content-wrapper" class="error snackbar">
          <p data-test-id="content" class="content">content</p><button class="iconButton" type="button">
            <phxcircle class="icon" aria-hidden="true" size="20" weight="duotone"></phxcircle><span class="text visually-hidden"> Clear snack </span>
          </button>
        </div>
      </div>
    </transition-group-stub>
  </div>
</transition-stub>"
`;

exports[`SnackBar > when snacks values are set > when value is { id: 'id', content: 'content', type: 'info' } > matches the snapshot 1`] = `
"<transition-stub name="snackbar" appear="false" persisted="false" css="true">
  <div class="wrapper">
    <transition-group-stub name="snackbar" appear="false" persisted="false" css="true">
      <div>
        <div data-test-id="content-wrapper" class="info snackbar">
          <p data-test-id="content" class="content">content</p><button class="iconButton" type="button">
            <phxcircle class="icon" aria-hidden="true" size="20" weight="duotone"></phxcircle><span class="text visually-hidden"> Clear snack </span>
          </button>
        </div>
      </div>
    </transition-group-stub>
  </div>
</transition-stub>"
`;

exports[`SnackBar > when snacks values are set > when value is { id: 'id', content: 'content', type: 'success' } > matches the snapshot 1`] = `
"<transition-stub name="snackbar" appear="false" persisted="false" css="true">
  <div class="wrapper">
    <transition-group-stub name="snackbar" appear="false" persisted="false" css="true">
      <div>
        <div data-test-id="content-wrapper" class="success snackbar">
          <p data-test-id="content" class="content">content</p><button class="iconButton" type="button">
            <phxcircle class="icon" aria-hidden="true" size="20" weight="duotone"></phxcircle><span class="text visually-hidden"> Clear snack </span>
          </button>
        </div>
      </div>
    </transition-group-stub>
  </div>
</transition-stub>"
`;
